akka-remote/src/main/scala/akka/remote/artery/Association.scala

```diff
@@ -467,22 +483,61 @@ private[remote] class Association(
 
   }
 
-  private def cancelIdleTimer(): Unit = {
-    val current = idle.get
+  private def cancelQuarantinedIdleTimer(): Unit = {
+    val current = quarantinedIdleTask.get
     current.foreach(_.cancel())
-    idle.compareAndSet(current, None)
+    quarantinedIdleTask.compareAndSet(current, None)
   }
 
-  private def startIdleTimer(): Unit = {
-    cancelIdleTimer()
-    idle.set(Some(transport.system.scheduler.scheduleOnce(advancedSettings.StopQuarantinedAfterIdle) {
+  private def startQuarantinedIdleTimer(): Unit = {
+    cancelQuarantinedIdleTimer()
+    quarantinedIdleTask.set(Some(transport.system.scheduler.scheduleOnce(advancedSettings.StopQuarantinedAfterIdle) {
       if (associationState.isQuarantined())
         streamMatValues.get.valuesIterator.foreach {
           case OutboundStreamMatValues(killSwitch, _) ⇒ killSwitch.abort(OutboundStreamStopSignal)
         }
     }(transport.system.dispatcher)))
   }
 
+  private def cancelIdleTimer(): Unit = {
+    val current = idleTask.get
+    current.foreach(_.cancel())
+    idleTask.compareAndSet(current, None)
+  }
+
+  private def startIdleTimer(): Unit = {
+    cancelIdleTimer()
+    val interval = settings.Advanced.StopIdleOutboundAfter / 2
+    val stopIdleOutboundAfterMillis = settings.Advanced.StopIdleOutboundAfter.toMillis
+    val task: Cancellable = transport.system.scheduler.schedule(interval, interval) {
+      if (System.currentTimeMillis() - associationState.lastUsedTimestamp.get >= stopIdleOutboundAfterMillis) {
+        var allStopped = true
+        streamMatValues.get.foreach {
+          case (queueIndex, OutboundStreamMatValues(killSwitch, _)) ⇒
+            if (isStreamActive(queueIndex)) {
+              if (queueIndex != ControlQueueIndex || associationState.pendingSystemMessagesCount.get == 0) {
+                log.debug("Stopping idle outbound stream [{}] to [{}]", queueIndex, remoteAddress)
+                killSwitch.abort(OutboundStreamStopSignal)
```

Comment:
I would also expect a removal of the association if all the streams are terminated

Reply:
I don't think we can avoid some race conditions when completely stopping/removing and those can result in lost messages. For ordinary messages that should be avoided but no catastrophe since we don't guarantee delivery (we promise best effort though). For system messages it is catastrophe since we must then quarantine the association. Therefore I would like to delay such removal so that it is more likely that no more messages will be sent to it.

That would of course be configurable and you could use a more aggressive approach.

In your case, do the connections become idle because the destination is shutdown and no more messages will be sent to them, or what is the message traffic pattern?

Reply:
That's probably the most pragmatic way to resolve the tradeoff between complexity of the solution and the risk to lose system messages. It will work just fine for me. Shutdown of the remote is possibly a cause for idle connections, but exploring the network for alternative neighbours is the more likely source in my case.